api/ratelimit/v1alpha1/ratelimit_types.go

```diff
@@ -0,0 +1,149 @@
+}
+
+// RatelimitSpec defines the desired state of Ratelimit
+type RatelimitSpec struct {
+	// Type decides the scope for the Ratelimits.
+	Type RatelimitType `json:"type"`
+	// Rules are a list of Ratelimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RatelimitRule `json:"rules"`
+}
+
+// RatelimitType specifies the types of Ratelimiting.
+// Valid RatelimitType values are:
+//
+// * "Local"
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Local;Global
+type RatelimitType string
+
+const (
+	// In this mode, the ratelimits are applied per Envoy proxy instance.
+	RatelimitTypeLocal RatelimitType = "Local"
+
+	// In this mode, the ratelimits are applied across all Envoy proxy instances.
+	RatelimitTypeGlobal RatelimitType = "Global"
+)
+
+// RatelimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RatelimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RatelimitMatch `json:"matches,omitempty"`
+	Limit   RatelimitValue   `json:"limit"`
+}
+
+// RatelimitMatch specifies the attributes within the traffic flow that can
+// be matched on.
+type RatelimitMatch struct {
+	// ClientAddress defines the semantics for matching on the IP Address
+	// of the client making the request.
+	// +optional
+	ClientAddress *ClientAddressMatch `json:"clientAddress,omitempty"`
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []HeaderMatch `json:"headers,omitempty"`
+}
+
+// ClientAddressMatch defines the match attributes based on the source IP Address
+// originated from the client.
+type ClientAddressMatch struct {
+	// Cidr specifies the subnet range of source IP addresses
+	// to be matched on.
+	// Not setting this field, implies matching on all source/client IP addresses.
+	// +optional
+	Cidr *string `json:"cidr,omitempty"`
+}
+
+// HeaderMatch defines the match attributes within the HTTP Headers of the request.
+type HeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name of the HTTP header.
+	Name string `json:"name"`
+
+	// Value within the HTTP header.
+	// Not setting this field, implies matching on all unique values within the header.
+	// +optional
+	Value *string `json:"value,omitempty"`
+}
+
+// HeaderMatchType specifies the semantics of how HTTP header values should be
+// compared. Valid HeaderMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
+//
+// +kubebuilder:validation:Enum=Exact;RegularExpression
+type HeaderMatchType string
+
+// HeaderMatchType constants.
+const (
+	HeaderMatchExact             HeaderMatchType = "Exact"
+	HeaderMatchRegularExpression HeaderMatchType = "RegularExpression"
+)
+
+// RatelimitValue defines the limits for ratelimiting.
+type RatelimitValue struct {
```

Comment:
how would this apply to `LocalRateLimit`?

Reply:
are you referring to what the implementation would look like ? here's one proposed solution - this could be converted internally into a token bucket config

Reply:
yes, with a converted example will be helpful.

Reply:
would prefer to keep any implementation detail out of this PR, so reviewers can focus on UX :)